### DISCUSSION TITLE
Create user by extending with go

### DISCUSSION THREAD

**Discussion by Clovis-krz** [pocketbase/pocketbase#3747] [Q&A] [answered]
Hi i am trying to create new users with fake data but for some reason the transaction says its successful but the user does not appear in DB. I do something like so : 
```
user_collection, err := app.Dao().FindCollectionByNameOrId("users")
	if err == nil {
		user_record := models.NewRecord(user_collection)
		user_form := forms.NewRecordUpsert(app, user_record)
		user_form.LoadData(map[string]any{
			"username":        gofakeit.Username(),
			"email":           gofakeit.Email(),
			"lastname":        gofakeit.LastName(),
			"firstname":       gofakeit.FirstName(),
			"password":        "lol",
			"passwordConfirm": "lol",
		})
		user_form.Verified = true
		if user_form.Submit(); err == nil {
			log.Println("insert " + user_record.Id + " to DB\n")
		} else {
			log.Println(err.Error())
		}
	}
```

**Comment by ganigeorgiev** [comment]
The above submit error check is missing an error assignment for the submit (aka. `err := user_form.Submit()`).

If that doesn't help, please provide a more detailed code sample so that I can try to reproduce it.

**Reply by ganigeorgiev** [reply]
_Also if you are running the above inside a db transaction (aka. `Dao().RunInTransaction(...)`), then you'll have to explicitly change the form's dao to the transaction one with [`user_form.SetDao(txDao)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.4/forms#RecordUpsert.SetDao)._

**Comment by Clovis-krz** [comment]
Thanks for your reply,
here is the output in the console of the code : ``` 2023/11/14 11:14:10 insert  to DB ``` . There is no error, but it still does not register the new user and the id of the user is not displayed in the output. Is there any other method i should use to create a user ? I am not running it inside a DB transaction

**Reply by ganigeorgiev** [reply]
That doesn't really help to understand what is going on.
If you are running the above code as it is, then you are NOT checking for the submit error because it is missing an assignment. `err` in the above is the result of the first `FindCollectionByNameOrId` call, not of the submit. What I mean by this is that `Submit()` can fail and still you'll print the sucess log.

If you are still not able to find the culprit, please create a **minimal** reproducible repo or at least provide a more detailed code sample to understand what is getting executed and when.

**Comment by Clovis-krz** [comment]
In users package :
```
package users

import (
	"fmt"
	"log"

	"github.com/brianvoe/gofakeit/v6"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/forms"
	"github.com/pocketbase/pocketbase/models"
)

func CreateUser(app *pocketbase.PocketBase) {
	user_collection, err := app.Dao().FindCollectionByNameOrId("users")
	if err == nil {
		user_record := models.NewRecord(user_collection)
		user_form := forms.NewRecordUpsert(app, user_record)
		user_form.LoadData(map[string]any{
			"username":  gofakeit.Username(),
			"email":     gofakeit.Email(),
			"lastname":  gofakeit.LastName(),
			"firstname": gofakeit.FirstName(),
		})
		user_record.SetPassword("lolipop")
		user_record.ValidatePassword("lolipop")
		user_record.PasswordHash()
		if user_form.Submit(); err != nil {
			log.Println(err.Error())
		} else {
			log.Println("insert " + user_record.Id + " to DB\n")
		}
	} else {
		log.Println(err.Error())
	}
}
```
In main : 
```
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"pocketbase/users"

	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/cron"
)

func main() {
	app := pocketbase.New()

	c := cron.New()
	c.MustAdd("FactoryUserScheduler", "* * * * *", func() {
		users.CreateUser(app)
	})
	c.Start()

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
To try the function i just added it as a contask every minute in the main for now. I hope this helps. If i still dont get what you want, can just please show me how to create a basic user in extending go ? Thanks again a lot for your help

**Reply by ganigeorgiev** [reply, marked as answer]
Replace:
```go
if user_form.Submit(); err != nil {
```

with:

```go
if err := user_form.Submit(); err != nil {
```
_(notice the `err` assignment)_


Also if you use the form to create the record, you don't have to manually call `SetPassword`, `ValidatePassword`, etc.

**Reply by Clovis-krz** [reply]
How yeah I understand know, sorry idk how i missed that ! Well it displayed the errors with password requirements and I fixed it ! Works perfectly. Thanks again a lot for your help and have a good day !
